text/0000-portability-lint.md

```diff
@@ -0,0 +1,574 @@
+    let fd = File::open("foo.txt").unwrap().as_raw_fd();
+}
+
+#[cfg(windows)]
+fn foo_windows() {
+    // No warning: we're within code that assumes `windows`
+    let handle = File::open("foo.txt").unwrap().as_raw_handle();
+}
+
+#[cfg(linux)]
+fn foo_linux() {
+    // No warning: we're within code that assumes `linux`, which implies `unix`
+    let fd = File::open("foo.txt").unwrap().as_raw_fd();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Code that provides a cross-platform abstraction
+////////////////////////////////////////////////////////////////////////////////
+
+// No `cfg` label here; it's a cross-platform function.
+fn cross_platform() {
+    // No warning emitted, since handling both `unix` and `windows` means
+    // this code works for all mainstream platforms.
+    match_cfg! {
+        unix => foo_unix(),
+        windows => foo_windows(),
+    }
+}
+```
+
+As with many lints, the portability lint is *best effort*: it is not required to
+provide airtight guarantees about portability. In particular, the proposed
+definition in this RFC ignores trait implementations and dispatch. In practice,
+this hole shouldn't be too significant: it's pretty rare to implement a trait
+conditionally based on platform information, and the goal of this RFC is to
+*guide* code toward mainstream compatibility without imposing a heavy burden.
+
+With that overview in mind, let's dig into the details.
+
+## The lint definition
+
+The lint is structured somewhat akin to a type and effect system: roughly
+speaking, items that are labeled with a given `cfg` assumption can only be used
+within code making that same `cfg` assumption.
+
+More precisely, each item has a *portability*, consisting of all the
+lexically-nested uses of `cfg`, including those expressed via `match_cfg`, but
+excluding mentions of Cargo features. If there are multiple uses of `cfg`, the
+portability is taken to be their *conjunction*:
+
+```rust
+#[cfg(unix)]
+mod foo {
+    #[cfg(target_pointer_width = "32")]
+    fn bar() {
+        // the portability of `bar` is `all(unix, target_pointer_width = "32")`
+    }
+}
+```
+
+The lint is then straightforward to define at a high level: it walks over item
+definitions and checks that the item's portability is *narrower* than the
+portability of items it references or invokes. For example, `bar` in the above
+could invoke an item with portability `unix` and/or `target_pointer_width =
+"32"`, but not one with portability `linux`.
+
+To fully define the lint, though, we need to give more details about what
+"narrower" means, and how referenced item portability is determined.
+
+### Comparing portabilities
+
+**What does it mean for a portability to be narrower?** In general, portability
+is a logical expression, using the operators `all`, `any`, `not` on top of
+primitive expressions like `unix`. Portability `P` is narrower than portability
+`Q` if `P` *implies* `Q` as a logic formula.
```

Comment:
If attribute values are extended to support basic arithmetic operations in the future (to support version requirements like `#[cfg(_MSC_VER >= 1600)]` and generally catch up with configuration abilities of C preprocessor), will it create additional problems for this scheme?